lib/jxl/enc_heuristics.cc

```diff
@@ -327,67 +373,326 @@ static void DownsampleImage2_Sharper(const ImageF& input, const ImageF& mask,
   }
 }
 
-static void GetMin2(const float v, float& min1, float& min2) {
-  if (v < min2) {
-    if (v < min1) {
-      min2 = min1;
-      min1 = v;
-    } else {
-      min2 = v;
+void DownsampleImage2_Sharper(Image3F* opsin) {
+  // Allocate extra space to avoid a reallocation when padding.
+  Image3F downsampled(DivCeil(opsin->xsize(), 2) + kBlockDim,
+                      DivCeil(opsin->ysize(), 2) + kBlockDim);
+  downsampled.ShrinkTo(downsampled.xsize() - kBlockDim,
+                       downsampled.ysize() - kBlockDim);
+
+  for (size_t c = 0; c < 3; c++) {
+    DownsampleImage2_Sharper(opsin->Plane(c), &downsampled.Plane(c));
+  }
+  *opsin = std::move(downsampled);
+}
+
+// The default upsampling kernels used by Upsampler in the decoder.
+static const constexpr int64_t kSize = 5;
+
+static const float kernel00[25] = {
+    -0.017162, -0.034523, -0.040222, -0.029210, -0.006246, -0.034523, 0.141111,
+    0.288968,  0.002787,  -0.016103, -0.040222, 0.288968,  0.566616,  0.037776,
+    -0.019867, -0.029210, 0.002787,  0.037776,  -0.031447, -0.011851, -0.006246,
+    -0.016103, -0.019867, -0.011851, -0.002135,
```

Comment:
shouldn't these kernels sum to 1.0

Reply:
They sum to 1.000001 for me, which sum do you get?